internal/weapons/sword/foliar/foliar.go

```diff
@@ -0,0 +1,88 @@
+package foliar
+
+import (
+	"fmt"
+
+	"github.com/genshinsim/gcsim/pkg/core"
+	"github.com/genshinsim/gcsim/pkg/core/attributes"
+	"github.com/genshinsim/gcsim/pkg/core/combat"
+	"github.com/genshinsim/gcsim/pkg/core/event"
+	"github.com/genshinsim/gcsim/pkg/core/glog"
+	"github.com/genshinsim/gcsim/pkg/core/keys"
+	"github.com/genshinsim/gcsim/pkg/core/player/character"
+	"github.com/genshinsim/gcsim/pkg/core/player/weapon"
+	"github.com/genshinsim/gcsim/pkg/modifier"
+)
+
+const (
+	buffKey = "foliar-whitemoon-bristle"
+	icdKey  = "foliar-icd"
+)
+
+func init() {
+	core.RegisterWeaponFunc(keys.FoliarIncision, NewWeapon)
+}
+
+type Weapon struct {
+	Index int
+}
+
+func (w *Weapon) SetIndex(idx int) { w.Index = idx }
+func (w *Weapon) Init() error      { return nil }
+
+func NewWeapon(c *core.Core, char *character.CharWrapper, p weapon.WeaponProfile) (weapon.Weapon, error) {
+	//CRIT Rate is increased by 4%. 
+	//After Normal Attacks deal Elemental DMG, the Foliar Incision effect will be obtained, 
+	//increasing DMG dealt by Normal Attacks and Elemental Skills by 120% of Elemental Mastery. 
+	//This effect will disappear after 28 DMG instances or 12s. You can obtain Foliar Incision once every 12s.
+	w := &Weapon{}
+	r := p.Refine
+
+	m := make([]float64, attributes.EndStatType)
+	m[attributes.CR] = 0.03 + float64(r)*0.01
+	char.AddStatMod(character.StatMod{
+		Base:         modifier.NewBase("foliar-crit-rate", -1),
+		AffectedStat: attributes.NoStat,
+		Amount: func() ([]float64, bool) {
+			return m, true
+		},
+	})
+
+	caBoost := 0.09 + 0.3*float64(r)
+	procCount := 0
+	c.Events.Subscribe(event.OnEnemyHit, func(args ...interface{}) bool {
```

Comment:
I think it would be cleaner to separate the activation of the effect and the actual effect itself. On that split you can then make the subscription that does the activation happen on OnEnemyDamage instead of OnEnemyHit. The flat dmg should stay on OnEnemyHit.